src/syntax/parser.rs

```diff
@@ -34,37 +34,34 @@ fn debunt(val: u64) -> String {
     return a;
 }
 
-impl<TokType: Tokenizer + Iterator<Item=Tokens>> Parser<TokType> {
-    pub fn new(tokenizer: TokType) -> Parser<TokType> {
+impl <I: Iterator<Item=Tokens>> Parser<I> {
+    pub fn new(token_stream: I) -> Parser<I> {
         Parser {
-            lexer: tokenizer,
+            token_stream: token_stream.peekable(),
             ast_root: ExprWrapper::default(Expr::NoOp),
             indent_level: 0,
             valid_ast: true,
-            preview_token: Some(Indent(0)),
+            preview_token: Some(Some(Indent(0))),
             block_status: BlockStatus::Out,
             between_brackets: false,
             last_depth: None,
         }
     }
 
-    /// Consume the next `Token` from the lexer
+    /// Consume the next `Token` from the token_stream
     /// - Ignores `Comment`s entirely
     /// - Smartly handlers `Indent`s by:
     ///    - When in blocks ignores them
     ///    - Ensures correct indentation size, then gets the next token
-    fn _next_token(&mut self, allow_any: bool) -> Tokens {
+    fn _next_token(&mut self, allow_any: bool) -> Option<Tokens> {
         loop {
-            let result = match self.preview_token.take() {
-                Some(tok) => tok,
-                None => match self.lexer.next() {
-                    Some(t) => t,
-                    None => Tokens::EOF,
-                },
+            let token = match self.preview_token.take() {
```

Comment:
I think this could be simplified to `let token = self.preview_token.take().unwrap_or_else(|| self.token_stream.next())`

Reply:
fair enough